drivers/vultr/vultr.go

```diff
@@ -0,0 +1,517 @@
+		if strings.Contains(err.Error(), "Invalid SSH Key") {
+			log.Infof("SSH key doesn't exist, assuming it is already deleted")
+		} else {
+			return err
+		}
+	}
+	if d.scriptID != 0 {
+		<-d.bucket.SpendToken(1)
+		if err := client.DeleteStartupScript(string(d.scriptID)); err != nil {
+			if strings.Contains(err.Error(), "Check SCRIPTID") {
+				log.Infof("Script doesn't exist, assuming it is already deleted")
+			} else {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (d *Driver) Restart() error {
+	if vmState, err := d.GetState(); err != nil {
+		return err
+	} else if vmState == state.Stopped {
+		log.Infof("Host is already stopped, use start command to run it")
+		return nil
+	}
+	log.Debugf("restarting %s", d.MachineName)
+	<-d.bucket.SpendToken(1)
+	return d.getClient().RebootServer(d.MachineID)
+}
+
+func (d *Driver) Kill() error {
+	if vmState, err := d.GetState(); err != nil {
+		return err
+	} else if vmState == state.Stopped {
+		log.Infof("Host is already stopped")
+		return nil
+	}
+	log.Debugf("killing %s", d.MachineName)
+	<-d.bucket.SpendToken(1)
+	return d.getClient().HaltServer(d.MachineID)
+}
+
+func (d *Driver) getClient() *vultr.Client {
+	return vultr.NewClient(d.APIKey, nil)
+}
+
+func (d *Driver) sshKeyPath() string {
+	return filepath.Join(d.storePath, "id_rsa")
+}
+
+func (d *Driver) publicSSHKeyPath() string {
+	return d.sshKeyPath() + ".pub"
+}
+
+func (d *Driver) instanceIsRunning() bool {
+	st, err := d.GetState()
+	if err != nil {
+		log.Debug(err)
+	}
+	if st == state.Running {
+		return true
+	}
+	log.Debug("VPS not yet started")
+	return false
+}
+
+func (d *Driver) validateApiCredentials() error {
+	_, err := d.getClient().GetAccountInfo()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (d *Driver) validateRegion() error {
+	<-d.bucket.SpendToken(1)
+	regions, err := d.getClient().GetRegions()
+	if err != nil {
+		return err
+	}
+	for _, region := range regions {
+		if region.ID == d.RegionID {
+			return nil
+		}
+	}
+	return fmt.Errorf("Region ID %d is invalid", d.RegionID)
+}
+
+func (d *Driver) validatePlan() error {
+	<-d.bucket.SpendToken(1)
+	plans, err := d.getClient().GetAvailablePlansForRegion(d.RegionID)
+	if err != nil {
+		return err
+	}
+	for _, v := range plans {
+		if v == d.PlanID {
+			return nil
+		}
+	}
+	return fmt.Errorf("Plan ID %d not available in this region. Available plans: %v", d.PlanID, plans)
+}
+
+// RancherOS - Create iPXE boot script
+func (d *Driver) createBootScript() error {
+	content := `#!ipxe
+set base-url https://releases.rancher.com/os/latest
+kernel ${base-url}/vmlinuz rancher.state.autoformat=[/dev/vda] rancher.cloud_init.datasources=[ec2]
+initrd ${base-url}/initrd
+boot`
+	<-d.bucket.SpendToken(1)
+	script, err := d.getClient().CreateStartupScript(d.MachineName, content, "pxe")
+	if err != nil {
+		return err
+	}
+	d.scriptID, err = strconv.Atoi(script.ID)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
```

Comment:
@hairyhenderson FYI, these two functions have been added after you last took a look (to allow booting RancherOS).